src/parser.rs

```diff
@@ -184,11 +176,11 @@ where
     }
 
     // FIXME(eddyb) safeguard this against misuse.
-    pub fn forest_add_split(&mut self, kind: P, left: Node<'i, P>) {
+    pub fn forest_add_split(&mut self, kind: G::NodeKind, left: Node<'i, G>) {
```

Comment:
Btw, this method largely has the same structure as the one above aside from the first statement and the argument to `.insert` -- refactor?

Reply:
These are intentionally kept separate (tempted to rename `split`/`choice` to `product`/`sum`, if that helps).